### PR TITLE
Add endpoint to notify unverified pubs they have money waiting

### LINKED ISSUE #375: [WIP] Email - Threshold met for site operator
**Conditions :** 
A domain exists which has acquired $10 worth of BAT and has thus far not started verification.

**Frequency :** 
Monthly until verification is complete.

**Subject :** 
 Money is waiting for <_domain_>, verify your site with Brave to claim your funds!

**Body :** 

Salutation : 
Hi!

Paragraph : 
Your site has accumulated quite a bit of fundage and we’d love to hand it over.    Please visit https://publishers.basicattentiontoken.org/ to start the verification process.  There are no fees, no gimmicks, we just want you to get paid and keep making awesome content.

Call-to-Action Button Text : 
Verify My Site

Post-script : 
Got questions? Visit https://basicattentiontoken.org/ to learn more about how the micropayments work.

We look forward to having you join the thousands of domains and YouTube channels already receiving funds from their audience.

### PR DESCRIPTION
**Pull request by nvonpentz**
A `POST` to `/api/publishers/notify_unverified` with the params:
````
publisher_type: "domain",
params: {
    domain: "<domain>"
}
````
Will scrape the WHOIS information for `<domain>` and send a notification email to each distinct valid email address found.  Currently, `publisher_type` can only be `"domain"` until we determine a solution for #372.  I chose this format for the params above because it matches those in the existing `PublisherNotifier` service class for verified publisher notifications.

I chose not to use the existing `PublisherNotifier` service class for this endpoint because that endpoint required an existing publisher record.  We would have to create a publisher record for each valid email address found for the unverified publisher, or refactor the `PublisherNotifier` class to not take a publisher record at all.  The requirements of these two seem different enough that they require their own endpoint and service class.

If we can't scrape any valid emails, no notification emails are sent and we return a 500.  If the params in the POST are malformed or incomplete, no emails are sent and we return a 400.

This is the notification email:
image

Resolves #375.
References #372

Submitter Checklist:
- [x] Submitted a [ticket](https://github.com/brave-intl/publishers/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Added/updated tests for this change (for new code or code which already has tests).
- [x] Tagged reviewers.

Test Plan:
Because the 'whois' gem does not support mocking, I separated the whois lookup into a service, `GetWhoisEmailsForDomain`.  The tests in `Api::PublishersControllerTest` mock the response from this service and perform some validations to make sure we do not send duplicate emails, or attempt to send to malformed email addresses.

There are also test for both new emails, `unverified_reached_threshold` and `unverified_reached_threshold_internal`.

Reviewer Checklist:

Tests
- [ ] Adequate test coverage exists to prevent regressions

Security:
- [ ] No raw SQL -- Always prefer ActiveRecord query helpers ([more info on StackOverflow](https://stackoverflow.com/questions/41410752/rails-5-sql-injection#41452695))
- [ ] XSS is mitigated -- Avoid `html_safe` and `raw`; escape untrusted content from users and 3rd party APIs ([Rails XSS guide](https://brakemanpro.com/2017/09/08/cross-site-scripting-in-rails) and [OWASP XSS Prevention Cheat Sheet](https://www.owasp.org/index.php/XSS_(Cross_Site_Scripting)_Prevention_Cheat_Sheet))

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -80,6 +80,11 @@ gem "clipboard-rails"
 # One-time passwords for 2fa
 gem "rotp", "~> 3.3"
 
+# WHOIS lookup for unverified publishers
+gem "whois", "~> 4.0"
+
+gem 'whois-parser', '~> 1.0'
```
Comment (ayumi) [inline Gemfile]: minor: prefer `"` for this project

File: Gemfile [hunk 2]
```diff
@@ -80,6 +80,11 @@ gem "clipboard-rails"
 # One-time passwords for 2fa
 gem "rotp", "~> 3.3"
 
+# WHOIS lookup for unverified publishers
+gem "whois", "~> 4.0"
```
Comment (ayumi) [inline Gemfile]: could you update `Gemfile.lock` as a result of the new gems?

Comment (ayumi) [reply Gemfile]: also, you could probably do `require: false` for this (then require it manually)

File: app/services/publisher_notifier_unverified.rb [hunk 3]
```diff
@@ -0,0 +1,67 @@
+class PublisherNotifierUnverified < BaseService
+  attr_reader :notification_params, :notification_type
+
+  PUBLISHER_TYPES = %w(youtube, domain)
```
Comment (ayumi) [inline app/services/publisher_notifier_unverified.rb]: let's remove the youtube references and require that the type be `domain` for now.  we can add a comment up top noting TODO: youtube

File: config/locales/en.yml [hunk 4]
```diff
@@ -216,14 +216,26 @@ en:
...
       erroneous_email_help: "If you did not make this request, you can ignore this email or contact us at "
       terms_of_service: "Terms of Service"
       medium: "Medium"
       twitter: "Twitter"
       reddit: "Reddit"
       rocketchat: "Rocket.Chat"
-
+    
+    unverified_domain_reached_threshold:
+      subject: "Money is waiting for [%{publication_title}], verify your site with Brave to claim your funds!"
```
Comment (ayumi) [inline config/locales/en.yml]: maybe we don't need the `[]`.. they look a little funny like it's a template email

**Comment by ayumi** [comment]
needs merge conflict resolution :)

**Comment by nvonpentz** [comment]
I settled on a POST to `/api/publishers/[brave_published_id]/notifications_unverified` with the params:
```
publisher_type: ["domain" | "youtube"],
publisher_id: ["example.com" | "UCtsfHRe2WQnkNH5WYJWL-Yw"]
```

Youtube is currently not implemented and will throw an error.  This is ready for review.
cc @mrose17

File: app/services/publisher_notifier_unverified.rb [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+    ensure_params_exist
+
+    if !PUBLISHER_TYPES.include?(@publisher_type)
+      raise InvalidPublisherTypeError.new("#{@publisher_type} is an invalid publisher type")
+    end
+  end
+
+  def perform
+    if @publisher_type == "domain"
```
Comment (ayumi) [inline app/services/publisher_notifier_unverified.rb]: minor: we plan on adding more platforms in the future, so a `case` statement might be more sense. however, we can do that later.

File: app/services/publisher_notifier_unverified.rb [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+
+    # send notification to valid emails
+    emails.each do |email|
+      PublisherMailer.unverified_domain_reached_threshold(domain, email).deliver_later
+      PublisherMailer.unverified_domain_reached_threshold_internal(domain, email).deliver_later
+    end
+  end
+
+  def perform_youtube
+    # TO DO
```
Comment (ayumi) [inline app/services/publisher_notifier_unverified.rb]: minor: might be better to leave out this out, or to have the method raise an error when called.

File: test/controllers/publishers_controller_test.rb [hunk 7]
```diff
@@ -357,7 +357,7 @@ def request_login_email(publisher:)
     # verify brave gets an internal email copy of confirmation email
     email = ActionMailer::Base.deliveries.find do |message|
       message.to == Rails.application.secrets[:internal_email]
-      message.subject == "<Internal> #{I18n.t('publisher_mailer.confirm_email_change.subject', publication_title: '')}"
```
Comment (ayumi) [inline test/controllers/publishers_controller_test.rb]: minor: unrelated change? tests seem ok though

File: app/services/publisher_notifier_unverified.rb [hunk 8]
```diff
@@ -0,0 +1,76 @@
+class PublisherNotifierUnverified < BaseService
+  attr_reader :notification_params, :notification_type
+
+  PUBLISHER_TYPES = %w(domain) # TODO: youtube
+
+  def initialize(publisher_type:, publisher_id: )
```
Comment (ayumi) [inline app/services/publisher_notifier_unverified.rb]: minor: unnecessary space after `publisher_id:`

File: app/services/publisher_notifier_unverified.rb [hunk 9]
```diff
@@ -0,0 +1,76 @@
+class PublisherNotifierUnverified < BaseService
+  attr_reader :notification_params, :notification_type
```
Comment (ayumi) [inline app/services/publisher_notifier_unverified.rb]: These are unused

File: app/services/publisher_notifier_unverified.rb [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+    ensure_params_exist
+
+    if !PUBLISHER_TYPES.include?(@publisher_type)
+      raise InvalidPublisherTypeError.new("#{@publisher_type} is an invalid publisher type")
+    end
+  end
+
+  def perform
+    if @publisher_type == "domain"
+      return perform_domain
```
Comment (ayumi) [inline app/services/publisher_notifier_unverified.rb]: minor: `return` here is redundant

File: app/services/publisher_notifier_unverified.rb [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+    else
+      return perform_youtube
+    end
+  end
+
+  def perform_domain
+    domain = @publisher_id
+    contacts = GetWhoisEmailsForDomain.new(@domain).perform
```
Comment (ayumi) [inline app/services/publisher_notifier_unverified.rb]: `@domain` is undefined, did you mean `@publisher_id`?

File: test/controllers/publishers_controller_test.rb (hunk 7 above)
Comment (nvonpentz) [reply test/controllers/publishers_controller_test.rb]: This test actually started to fail. I suspect it was because we removed the brackets from the email templates.  This seemed to fix it though.

File: app/services/publisher_notifier_unverified.rb (hunk 6 above)
Comment (nvonpentz) [reply app/services/publisher_notifier_unverified.rb]: It should never be called since 'youtube' is not an accepted `PUBlSHER_TYPE`, but I will remove it until we implement it.

File: test/controllers/publishers_controller_test.rb (hunk 7 above)
Comment (mixonic) [reply test/controllers/publishers_controller_test.rb]: hm, I'm having to change this back to having a space in a current branch 🤔
